backends/bmv2/simple_switch/midend.cpp

```diff
@@ -126,10 +126,10 @@ SimpleSwitchMidEnd::SimpleSwitchMidEnd(CompilerOptions& options, std::ostream* o
             // control plane API, we remove them as well for P4-14 programs.
             isv1 ? new P4::RemoveUnusedActionParameters(&refMap) : nullptr,
             new P4::TypeChecking(&refMap, &typeMap),
-            options.loopsUnrolling ? new P4::ParsersUnroll(true, &refMap, &typeMap) : nullptr,
             new P4::MidEndLast(),
             evaluator,
             [this, evaluator]() { toplevel = evaluator->getToplevelBlock(); },
+            options.loopsUnrolling ? new P4::ParsersUnroll(true, &refMap, &typeMap) : nullptr,
```

Comment:
MidEndLast is supposed to be the last pass in the midend.
What was wrong with the previous position?

Reply:
We need to make lines 130,131 before new P4::ParsersUnroll. I have placed it at the end.

Reply:
Why?
Technically the result of evaluation is used after the midend to generate code. So if you perform additional transformations on the code after the evaluator, the subsequent pass may not work, since it operates on a different structure: toplevel no longer reflects the program structure.
Frankly, the evaluator pass is a mistake, but we don't have the cycles to remove it.
Loop unrolling should really be insertable almost anywhere. Why does it have to be last?
In fact, I would try to place it early, so other optimizations, like constant folding, can take advantage of your changes.